orm/model/ormtable/primary_key.go

```diff
@@ -221,4 +223,126 @@ func (p primaryKeyIndex) deleteByIterator(ctx context.Context, it Iterator) erro
 	return writer.Write()
 }
 
+func (t primaryKeyIndex) doSave(writer *batchIndexCommitmentWriter, message proto.Message, mode saveMode) error {
+	mref := message.ProtoReflect()
+	pkValues, pk, err := t.EncodeKeyFromMessage(mref)
+	if err != nil {
+		return err
+	}
+
+	existing := mref.New().Interface()
+	haveExisting, err := t.getByKeyBytes(writer, pk, pkValues, existing)
+	if err != nil {
+		return err
+	}
+
+	if !haveExisting {
+		existing = nil
+	}
+
+	err = t.doUpdateWithWriteBatch(writer, pkValues, pk, existing, message, mode)
+	if err != nil {
+		return err
+	}
+
+	return writer.Write()
+}
+
+func (p primaryKeyIndex) doUpdateWithWriteBatch(writer *batchIndexCommitmentWriter, pkValues []protoreflect.Value, pkBz []byte, existing, new protoreflect.ProtoMessage, mode saveMode) error {
+	mref := new.ProtoReflect()
+	if existing != nil {
+		if mode == saveModeInsert {
+			return ormerrors.PrimaryKeyConstraintViolation.Wrapf("%q:%+v", mref.Descriptor().FullName(), pkValues)
+		}
+
+		if hooks := writer.Hooks(); hooks != nil {
+			err := hooks.OnUpdate(existing, new)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		if mode == saveModeUpdate {
+			return ormerrors.NotFoundOnUpdate.Wrapf("%q", mref.Descriptor().FullName())
+		}
+
+		if hooks := writer.Hooks(); hooks != nil {
+			err := hooks.OnInsert(new)
+			if err != nil {
+				return err
```

Comment:
so hooks can stop the whole process?

Reply:
yes, I am going to separate before and after hooks though, and will put this in draft and add that in a separate PR